### ISSUE TITLE
File::Spec::abs2rel wrong result when path is ../something

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13635]
Migrated from [rt.perl.org#121360](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121360) (status was 'resolved')

Searchable as RT121360$

**Comment by p5pRT** [comment]
### From @jimav


This is a bug report for perl from james\_avera@&#8203;yahoo\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.



---
File&#8203;::Spec\->abs2rel\($path\, $base\) is supposed to allow 
$path and/or $base to be relative to the current working directory\,
and the pod says that if either are relative\, they are
converted to absolute using "rel2abs\(\)"\.

However a wrong result is returned on Linux if both $path and $base are
relative and $path is \.\./something\.

The problem does not occur if rel2abs\(\) is called explicitly first\,
so abs2rel must be doing something other than just calling
rel2abs on the arguments\.

\#\!/usr/bin/perl
use strict; use warnings;
use Cwd qw\(getcwd abs\_path\);
use File&#8203;::Spec\(\);

chdir "/etc" or die;
\#my $rel\_path  = "passwd";        \#works
\#my $rel\_path  = "\./passwd";      \#works
\#my $rel\_path  = "\.\./etc/passwd"; \#fails
my $rel\_path  = "\.\./dev/null";    \#fails
my $rel\_base = "init\.d";
my $abs\_path = File&#8203;::Spec\->rel2abs\($rel\_path\);
my $abs\_base = File&#8203;::Spec\->rel2abs\($rel\_base\);
warn "getcwd   = "\,getcwd\(\)\,"\\n";
warn "\\n";
warn "rel\_path = $rel\_path\\n";
warn "rel\_base = $rel\_base\\n";
warn "\\n";
warn "abs\_path = $abs\_path\\n";
warn "abs\_base = $abs\_base\\n";
warn "\\n";
if \(\-r $rel\_path && \-r $abs\_path && \[stat\($rel\_path\)\] ~~ \[stat\($abs\_path\)\]\) \{
&nbsp;  warn "rel\_path & abs\_path are valid and the same object\\n";
\} else \{die\}
warn "\\n";
if \(\-d $rel\_base && \-d $abs\_base && \[stat\($rel\_base\)\] ~~ \[stat\($abs\_base\)\]\) \{
&nbsp;  warn "rel\_base & abs\_base are valid and the same directory\\n";
\} else \{die\}
warn "\\n";
my $rr\_link = File&#8203;::Spec\->abs2rel\($rel\_path\, $rel\_base\);
my $ra\_link = File&#8203;::Spec\->abs2rel\($rel\_path\, $abs\_base\);
my $ar\_link = File&#8203;::Spec\->abs2rel\($abs\_path\, $rel\_base\);
my $aa\_link = File&#8203;::Spec\->abs2rel\($abs\_path\, $abs\_base\);
warn "abs2rel\(rel\_path\, rel\_base\) returned $rr\_link\\n";
warn "abs2rel\(rel\_path\, abs\_base\) returned $ra\_link\\n";
warn "abs2rel\(abs\_path\, rel\_base\) returned $ar\_link\\n";
warn "abs2rel\(abs\_path\, rel\_base\) returned $aa\_link\\n";
die "BUG" unless $rr\_link eq $ra\_link; \# dies here
die "BUG" unless $ar\_link eq $aa\_link;
die "BUG" unless $rr\_link eq $aa\_link;



---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=medium
&nbsp;    module=File&#8203;::Spec


---
Site configuration information for perl 5\.14\.2&#8203;:

Configured by Debian Project at Thu Jul 18 22&#8203;:04&#8203;:35 UTC 2013\.

Summary of my perl5 \(revision 5 version 14 subversion 2\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.2\.0\-37\-generic\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux roseapple 3\.2\.0\-37\-generic \#58\-ubuntu smp thu jan 24 15&#8203;:28&#8203;:10 utc 2013 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fstack\-protector \-\-param=ssp\-buffer\-size=4 \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.14 \-Darchlib=/usr/lib/perl/5\.14 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.14\.2 \-Dsitearch=/usr/local/lib/perl/5\.14\.2 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.14\.2 \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.14\.2
&nbsp;    gnulibc\_version='2\.17'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.14\.2&#8203;:
&nbsp;    /home/jima/lib/perl
&nbsp;    /home/jima/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/jima/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/jima/perl5/lib/perl5
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.14\.2
&nbsp;    /usr/local/share/perl/5\.14\.2
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.14
&nbsp;    /usr/share/perl/5\.14
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.14\.2&#8203;:
&nbsp;    HOME=/home/jima
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE=en\_US
&nbsp;    LD\_LIBRARY\_PATH=/home/jima/local/lib
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/jima/perl5/bin&#8203;:/home/jima/bin&#8203;:/home/jima/local/bin&#8203;:/home/jima/jima\_tools/x86\_64/bin&#8203;:/home/jima/jima\_tools/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/bin/X11&#8203;:/usr/local/bin&#8203;:/usr/lib/lightdm/lightdm&#8203;:/usr/local/sbin&#8203;:/usr/games&#8203;:/usr/local/games&#8203;:/usr/lib/jvm/java\-7\-oracle/bin&#8203;:/usr/lib/jvm/java\-7\-oracle/db/bin&#8203;:/usr/lib/jvm/java\-7\-oracle/jre/bin&#8203;:\.
&nbsp;    PERL5LIB=/home/jima/lib/perl&#8203;:/home/jima/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi&#8203;:/home/jima/perl5/lib/perl5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_LOCAL\_LIB\_ROOT=/home/jima/perl5
&nbsp;    PERL\_MB\_OPT=\-\-install\_base /home/jima/perl5
&nbsp;    PERL\_MM\_OPT=INSTALL\_BASE=/home/jima/perl5
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @ikegami

What output did you get?

I get the following and it doesn't die \(5\.18\.0 on Linux\)

getcwd   = /etc

rel\_path = \.\./dev/null
rel\_base = init\.d

abs\_path = /etc/\.\./dev/null
abs\_base = /etc/init\.d

rel\_path & abs\_path are valid and the same object

rel\_base & abs\_base are valid and the same directory

abs2rel\(rel\_path\, rel\_base\) returned \.\./\.\./dev/null
abs2rel\(rel\_path\, abs\_base\) returned \.\./\.\./dev/null
abs2rel\(abs\_path\, rel\_base\) returned \.\./\.\./dev/null
abs2rel\(abs\_path\, rel\_base\) returned \.\./\.\./dev/null

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jimav

 > What output did you get?

PERL\_VERSION=v5\.14\.2
File&#8203;::Spec&#8203;::VERSION=3\.33
File&#8203;::Spec&#8203;::Unix&#8203;::VERSION=3\.33
getcwd   = /etc

rel\_path = \.\./dev/null
rel\_base = init\.d

abs\_path = /etc/\.\./dev/null
abs\_base = /etc/init\.d

rel\_path & abs\_path are valid and the same object

rel\_base & abs\_base are valid and the same directory

abs2rel\(rel\_path\, rel\_base\) returned \.\./dev/null
abs2rel\(rel\_path\, abs\_base\) returned \.\./\.\./dev/null
abs2rel\(abs\_path\, rel\_base\) returned \.\./\.\./dev/null
abs2rel\(abs\_path\, abs\_base\) returned \.\./\.\./dev/null
BUG at /tmp/t3 line 43\.


So it looks like an old bug which has been fixed \(?\)

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Mar 03 18&#8203;:57&#8203;:11 2014\, jimav wrote&#8203;:
>  > What output did you get?
> 
> PERL\_VERSION=v5\.14\.2
> File&#8203;::Spec&#8203;::VERSION=3\.33
> File&#8203;::Spec&#8203;::Unix&#8203;::VERSION=3\.33
> getcwd   = /etc
> 
> rel\_path = \.\./dev/null
> rel\_base = init\.d
> 
> abs\_path = /etc/\.\./dev/null
> abs\_base = /etc/init\.d
> 
> rel\_path & abs\_path are valid and the same object
> 
> rel\_base & abs\_base are valid and the same directory
> 
> abs2rel\(rel\_path\, rel\_base\) returned \.\./dev/null
> abs2rel\(rel\_path\, abs\_base\) returned \.\./\.\./dev/null
> abs2rel\(abs\_path\, rel\_base\) returned \.\./\.\./dev/null
> abs2rel\(abs\_path\, abs\_base\) returned \.\./\.\./dev/null
> BUG at /tmp/t3 line 43\.
> 
> 
> So it looks like an old bug which has been fixed \(?\)
> 

I would like to move this ticket toward resolution\.  To do so\, I have created a new test file for dist/PathTools/t/\, attached\.  This file attempts to simulate the script provided earlier in this ticket in temporary directories rather in /etc \(which probably wouldn't be portable\)\.

That test program is attached in 121360\-Demonstrate\-that\-RT\-121360\-has\-been\-resolved\.patch\.  I've also pushed that for smoking to the smoke\-me/jkeenan/121360\-abs2rel branch\.

I suspect that before we apply this patch we should adapt it for running on VMS\, as we've done for similar test files in that distribution and others\.  I welcome suggestions for that\.  Please review\.

Thank you very much\.
Jim Keenan

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>121360-Demonstrate-that-RT-121360-has-been-resolved.patch</summary>

```diff
From fead07be34a654cf19530ea648df5ffe04cc2841 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Tue, 21 Oct 2014 19:29:33 -0400
Subject: [PATCH] Demonstrate that RT #121360 has been resolved.

Bug report filed by James Avera argued that "File::Spec->abs2rel($path, $base)
is supposed to allow $path and/or $base to be relative to the current working
directory, and the pod says that if either are relative, they are converted to
absolute using 'rel2abs()'."

The cases reported as failing were subsequently reported as passing.  Adding a
test file to confirm this.
---
 MANIFEST                   |   1 +
 dist/PathTools/t/abs2rel.t | 108 +++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 109 insertions(+)
 create mode 100644 dist/PathTools/t/abs2rel.t

diff --git a/MANIFEST b/MANIFEST
index 8fb37a6..1b5b7e6 100644
--- a/MANIFEST
+++ b/MANIFEST
@@ -3209,6 +3209,7 @@ dist/PathTools/lib/File/Spec/Unix.pm		portable operations on Unix file names
 dist/PathTools/lib/File/Spec/VMS.pm		portable operations on VMS file names
 dist/PathTools/lib/File/Spec/Win32.pm		portable operations on Win32 and NetWare file names
 dist/PathTools/Makefile.PL			makefile writer for Cwd
+dist/PathTools/t/abs2rel.t		See if File::Spec->abs2rel works
 dist/PathTools/t/crossplatform.t		See if File::Spec works crossplatform
 dist/PathTools/t/cwd.t			See if Cwd works
 dist/PathTools/t/Functions.t			See if File::Spec::Functions works
diff --git a/dist/PathTools/t/abs2rel.t b/dist/PathTools/t/abs2rel.t
new file mode 100644
index 0000000..5c5431a
--- /dev/null
+++ b/dist/PathTools/t/abs2rel.t
@@ -0,0 +1,108 @@
+#!/usr/bin/perl -w
+
+use strict;
+use Test::More;
+
+use Cwd qw(cwd getcwd abs_path);
+use File::Spec();
+use File::Temp qw(tempdir);
+use File::Path qw(make_path);
+use Data::Dumper;
+
+my $startdir = cwd();
+
+test_rel2abs( {
+    startdir        => $startdir,
+    first_sub_dir   => 'etc',
+    sub_sub_dir     => 'init.d',
+    first_file      => 'passwd',
+    second_sub_dir  => 'dev',
+    second_file     => 'null',
+} );
+
+test_rel2abs( {
+    startdir        => $startdir,
+    first_sub_dir   => 'etc',
+    sub_sub_dir     => 'init.d',
+    first_file      => './passwd',
+    second_sub_dir  => 'dev',
+    second_file     => 'null',
+} );
+
+test_rel2abs( {
+    startdir        => $startdir,
+    first_sub_dir   => 'etc',
+    sub_sub_dir     => 'init.d',
+    first_file      => '../etc/passwd',
+    second_sub_dir  => 'dev',
+    second_file     => 'null',
+} );
+
+test_rel2abs( {
+    startdir        => $startdir,
+    first_sub_dir   => 'etc',
+    sub_sub_dir     => 'init.d',
+    first_file      => '../dev/null',
+    second_sub_dir  => 'dev',
+    second_file     => 'null',
+} );
+
+sub test_rel2abs {
+    my $args = shift;
+    my $tdir = tempdir( CLEANUP => 1 );
+    chdir $tdir or die "Unable to change to $tdir: $!";
+    
+    my @subdirs = (
+        $args->{first_sub_dir},
+        File::Spec->catdir($args->{first_sub_dir},  $args->{sub_sub_dir}),
+        $args->{second_sub_dir}
+    );
+    make_path(@subdirs, { mode => 0711 })
+        or die "Unable to make_path: $!";
+    
+    open my $OUT2, '>',
+        File::Spec->catfile($args->{second_sub_dir}, $args->{second_file})
+        or die "Unable to open $args->{second_file} for writing: $!";
+    print $OUT2 "Attempting to resolve RT #121360\n";
+    close $OUT2 or die "Unable to close $args->{second_file} after writing: $!";
+
+    chdir $args->{first_sub_dir}
+        or die "Unable to change to '$args->{first_sub_dir}': $!";
+    open my $OUT1, '>', $args->{first_file}
+        or die "Unable to open $args->{first_file} for writing: $!";
+    print $OUT1 "Attempting to resolve RT #121360\n";
+    close $OUT1 or die "Unable to close $args->{first_file} after writing: $!";
+
+    my $rel_path = $args->{first_file};
+    my $rel_base = $args->{sub_sub_dir};
+    my $abs_path = File::Spec->rel2abs($rel_path);
+    my $abs_base = File::Spec->rel2abs($rel_base);
+    ok(-f $rel_path, "'$rel_path' is readable by effective uid/gid");
+    ok(-f $abs_path, "'$abs_path' is readable by effective uid/gid");
+    is_deeply(
+        [ (stat $rel_path)[0..5] ],
+        [ (stat $abs_path)[0..5] ],
+        "rel_path and abs_path stat same"
+    );
+    ok(-d $rel_base, "'$rel_base' is a directory");
+    ok(-d $abs_base, "'$abs_base' is a directory");
+    is_deeply(
+        [ (stat $rel_base)[0..5] ],
+        [ (stat $abs_base)[0..5] ],
+        "rel_base and abs_base stat same"
+    );
+    my $rr_link = File::Spec->abs2rel($rel_path, $rel_base);
+    my $ra_link = File::Spec->abs2rel($rel_path, $abs_base);
+    my $ar_link = File::Spec->abs2rel($abs_path, $rel_base);
+    my $aa_link = File::Spec->abs2rel($abs_path, $abs_base);
+    is($rr_link, $ra_link,
+        "rel_path-rel_base '$rr_link' = rel_path-abs_base '$ra_link'");
+    is($ar_link, $aa_link,
+        "abs_path-rel_base '$ar_link' = abs_path-abs_base '$aa_link'");
+    is($rr_link, $aa_link,
+        "rel_path-rel_base '$rr_link' = abs_path-abs_base '$aa_link'");
+
+    chdir $args->{startdir} or die "Unable to change back to $args->{startdir}: $!";
+}
+
+done_testing();
-- 
1.9.1


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Oct 21 17&#8203;:34&#8203;:56 2014\, jkeenan wrote&#8203;:
> That test program is attached in 121360\-Demonstrate\-that\-RT\-121360\-
> has\-been\-resolved\.patch\.  I've also pushed that for smoking to the
> smoke\-me/jkeenan/121360\-abs2rel branch\.
> 
> I suspect that before we apply this patch we should adapt it for
> running on VMS\, as we've done for similar test files in that
> distribution and others\.  I welcome suggestions for that\.  Please
> review\.

It looks fine to me\, except for the ‘use Data&#8203;::Dumper’\, which can just be deleted\, unless I missed something\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Oct 21 20&#8203;:38&#8203;:36 2014\, sprout wrote&#8203;:

> 
> It looks fine to me\, except for the ‘use Data&#8203;::Dumper’\, which can just
> be deleted\, unless I missed something\.

Applied to blead\, with that and whitespace corrections\, in commit ec9e47ebf20db99adb00e868eff8a50a70b15b71

I'm going to hold the ticket open for a while to see if Craig or someone else can make suggestions for adapting the test to VMS\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Oct 28 18&#8203;:27&#8203;:34 2014\, jkeenan wrote&#8203;:

> Applied to blead\, with that and whitespace corrections\, in commit
> ec9e47ebf20db99adb00e868eff8a50a70b15b71
> 
> I'm going to hold the ticket open for a while to see if Craig or
> someone else can make suggestions for adapting the test to VMS\.
> 

No such suggestions have been forthcoming since late last month\.  So I'm going to mark this ticket resolved\.  If these tests don't work on VMS or any other OS\, please open a new RT\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @craigberry

On Sat\, Nov 22\, 2014 at 3&#8203;:11 PM\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue Oct 28 18&#8203;:27&#8203;:34 2014\, jkeenan wrote&#8203;:
>
>> Applied to blead\, with that and whitespace corrections\, in commit
>> ec9e47ebf20db99adb00e868eff8a50a70b15b71
>>
>> I'm going to hold the ticket open for a while to see if Craig or
>> someone else can make suggestions for adapting the test to VMS\.
>>
>
> No such suggestions have been forthcoming since late last month\.  So I'm going to mark this ticket resolved\.  If these tests don't work on VMS or any other OS\, please open a new RT\.

I fixed it for VMS already&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/history/HEAD&#8203;:/dist/PathTools

It did identify some things that were broken in File&#8203;::Spec&#8203;::VMS\, so
it's a good test from that point of view\.  Sorry I forgot to update
the ticket \-\- I thought it had already been closed\.
